internal/executor/executor.go

```diff
@@ -294,32 +328,52 @@ func (executor *Executor) performStep(ctx context.Context, env map[string]string
 			signaledToExit = false
 		}
 	case *api.Command_BackgroundScriptInstruction:
-		cmd, logClient, err := executor.ExecuteScriptsAndStreamLogs(ctx, currentStep.Name, instruction.BackgroundScriptInstruction.Scripts, env)
+		cmd, err := executor.ExecuteScriptsAndStreamLogs(ctx, logUploader,
+			instruction.BackgroundScriptInstruction.Scripts, executor.env)
 		if err == nil {
 			executor.backgroundCommands = append(executor.backgroundCommands, CommandAndLogs{
 				Name: currentStep.Name,
 				Cmd:  cmd,
-				Logs: logClient,
+				Logs: logUploader,
 			})
 			log.Printf("Started execution of #%d background command %s\n", len(executor.backgroundCommands), currentStep.Name)
 			success = true
 		} else {
 			log.Printf("Failed to create command line for background command %s: %s\n", currentStep.Name, err)
-			_, _ = logClient.Write([]byte(fmt.Sprintf("Failed to create command line: %s", err)))
-			logClient.Finalize()
+			_, _ = logUploader.Write([]byte(fmt.Sprintf("Failed to create command line: %s", err)))
+			logUploader.Finalize()
 			success = false
 		}
 	case *api.Command_CacheInstruction:
-		success = executor.DownloadCache(ctx, currentStep.Name, executor.httpCacheHost, instruction.CacheInstruction, env)
+		success = executor.DownloadCache(ctx, logUploader, currentStep.Name, executor.httpCacheHost,
+			instruction.CacheInstruction, executor.env)
 	case *api.Command_UploadCacheInstruction:
-		success = executor.UploadCache(ctx, currentStep.Name, executor.httpCacheHost, instruction.UploadCacheInstruction, env)
+		success = executor.UploadCache(ctx, logUploader, currentStep.Name, executor.httpCacheHost,
+			instruction.UploadCacheInstruction, executor.env)
 	case *api.Command_ArtifactsInstruction:
-		success = executor.UploadArtifacts(ctx, currentStep.Name, instruction.ArtifactsInstruction, env)
+		success = executor.UploadArtifacts(ctx, logUploader, currentStep.Name,
+			instruction.ArtifactsInstruction, executor.env)
 	default:
 		log.Printf("Unsupported instruction %T", instruction)
 		success = false
 	}
 
+	cirrusEnvVariables, err := cirrusEnv.Consume()
```

Comment:
What do you think about moving it before execution and in case of an error write the error message to `logUploader`?

Reply:
It's already logged to `logUploader` here: https://github.com/cirruslabs/cirrus-ci-agent/pull/144/files#diff-7b2829bb086330a6edbf89c117b46087d74ebe7c7e17a7567d2f3b7a11241ed7R365.

And I don't quite understand why you want to move it before execution? Thie `Consume()` reads the `CIRRUS_ENV` file, which is only populated after execution 🤔

Reply:
I suggested to evaluate it before in case of my first comment with a single env file for a task.